File: fortran/common/Fibonacci.f90
!    File:    Fibonacci.f90
!    Author:  Marvin Smith
!    Date:    4/25/2015
!
!    Purpose: Contains some Fibonacci utilities written in Fortran.
!
!    Fibonacci Module
MODULE FibonacciModule

!  Define Everything As Public
PUBLIC

!  List of Procedures in the Fibonacci Module
CONTAINS

!   Recursive Fibonacci Function
RECURSIVE FUNCTION Fibonacci (n) result (fnum)
     integer, intent(in) :: n
     integer :: fnum

    !  Handle Exit Conditions
    if (n<2) then
        fnum = n
    
    !  Handle Normal Conditions
    else
        fnum = Fibonacci(n-1) + Fibonacci(n-2)

    endif
end function Fibonacci


!   Find the Max Fibonacci Number given Max Value
FUNCTION Find_Max_Fibonacci (max_value) result (Fmax)
    integer, intent(in) :: max_value
    integer :: Fmax, I, val


!  Start iterating
I = 2
Fmax = I
do

    !  Compute the Fibonacci Number
    val = Fibonacci(I)

    !  Check if we are at the max number
    if ( val >= max_value ) then
        EXIT
    endif

    !  Increment the Counter
    Fmax = I
    I = I + 1

END DO
END FUNCTION

!   End of the Module
END MODULE FibonacciModule

File: octave/common/Fibonacci.m
%    File:    Fibonacci.m
%    Author:  Marvin Smith
%    Date:    4/25/2015
%
%    Purpose:  Perform the fibonacci operation
%

function result = Fibonacci(F)

    persistent fib_table;

    %  Check if fib table is empty
    if isempty(fib_table)
        fib_table = cell(1,2);
        fib_table{1,1} = 1;
        fib_table{1,2} = 2;
    end

    % Check Exit Conditions
    if F < 3
        result = fib_table{1,F};

    %  Check if value already solved
    else if size(fib_table)(2) >= F && isnumeric(fib_table{1,F}) == 1 
        result = fib_table{1,F};
       
    %  Standard method
    else
        fib_table{1,F} = Fibonacci(F-1) + Fibonacci(F-2);
        result = fib_table{1,F};
    end

end

